File: docs/chapters/01-solid-and-core-engineering-principles/go/dependency_inversion_ports.go
// Scenario: order placement charges via PSP without importing Stripe SDK in domain
// Demonstrates: Dependency Inversion — OrderService depends on PaymentPort interface
// Trade-off: explicit error returns vs exceptions; same boundary as Java

package main

type ChargeRequest struct {
	OrderID      string
	AmountCents  int64
	Currency     string
}

type ChargeResult struct {
	ProviderRef string
	Success     bool
}

type PaymentPort interface {
	Charge(req ChargeRequest) (ChargeResult, error)
}

type OrderService struct{ payments PaymentPort }

func (s *OrderService) PlaceOrder(req ChargeRequest) (string, bool) {
	res, err := s.payments.Charge(req)
	if err != nil || !res.Success {
		return "", false
	}
	return res.ProviderRef, true
}

type stripeAdapter struct{}

func (stripeAdapter) Charge(req ChargeRequest) (ChargeResult, error) {
	return ChargeResult{ProviderRef: "ch_mock", Success: true}, nil
}

File: docs/chapters/01-solid-and-core-engineering-principles/java/DependencyInversionPorts.java
// Scenario: order placement charges via PSP without importing Stripe SDK in domain
// Demonstrates: Dependency Inversion — OrderService depends on PaymentPort
// Trade-off: extra interface vs testability and PSP swap

import java.util.Optional;

record ChargeRequest(String orderId, long amountCents, String currency) {}
record ChargeResult(String providerRef, boolean success) {}

interface PaymentPort {
    ChargeResult charge(ChargeRequest request);
}

final class OrderService {
    private final PaymentPort payments;
    OrderService(PaymentPort payments) { this.payments = payments; }

    Optional<String> placeOrder(ChargeRequest req) {
        ChargeResult res = payments.charge(req);
        return res.success() ? Optional.of(res.providerRef()) : Optional.empty();
    }
}

// Infrastructure adapter (would live in infra package in production)
final class StripeAdapter implements PaymentPort {
    public ChargeResult charge(ChargeRequest request) {
        // call Stripe SDK here
        return new ChargeResult("ch_mock", true);
    }
}
